File: euler-problems/problems1-10/problem5/main.dart
/**
 * 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
 * What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
 * Result: 232792560
 */


int gcd(int a, int b) => b == 0 ? a : gcd(b, a % b);

int lcm(int a, int b) => (a * b) ~/ gcd(a, b);

Iterable<int> int_range(int start, int end) sync* {
  for (int i = start; i < end; i++) {
    yield i;
  }
}

int solve() => int_range(1, 21).reduce(lcm);

main(List<String> arguments) {
  print(solve());
}

File: euler-problems/problems1-10/problem5/main.kt
/**
 * 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
 * What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
 * Result: 232792560
 */

tailrec fun gcd(a: Long, b: Long): Long = if (b == 0L) a else gcd(b, a % b)

fun lcm(a: Long, b: Long) = (a * b) / gcd(a, b)

fun solve() = (1L..20L).reduce(::lcm)

fun main(args: Array<String>) {
    solve().let(::println)
}